### PR TITLE
tweaking api and updating to swift 3

### PR DESCRIPTION
**Pull request by loganwright**
Some updates:
- Update to Swift 3 Syntax (2.2 compatible)
- Remove need for explicit wrapped type in generic declarations, ie: `Validated<String, EmptyString>` can now just be `Validated<EmptyString>`
- SPM Support

### CODE REVIEW

**Comment by loganwright** [comment]
I updated so it's also compatible w/ 2.2. It was pretty simple, so I felt like it was worth while. 

I bumped podspec etc., but you'll need to update tag on repo.

**Comment by Ben-G** [comment]
Hey @LoganWright, thanks a lot for this contribution!

Most of these changes look great! There's one change I think I'm not in agreement with:

> Remove need for explicit wrapped type in generic declarations, ie: Validated<String, EmptyString> can now just be Validated<EmptyString>

IMO it is better to be explicit here. One of the core ideas of Validated is to wrap an existing type with a validation; therefore I think it's useful to explicitly state that underlying type even though it might be slightly redundant sometimes.

Happy to hear your thoughts on this!

**Comment by loganwright** [comment]
Hi @Ben-G 

Thanks for your response, glad to hear some of these changes will be useful!

On the explicit input types, my thought process was based on this rule from the latest official Swift api design guidelines.

> Omit needless words.

I think since the Validator provides the information necessary, we can use that to make our code more concise. Particularly when composing w/ `And`, `Not`, etc. it can be nice to have a bit more concision.

Let me know what you decide 👍

**Comment by Ben-G** [comment]
Hah, that's coincidental, I considered quoting the Swift API Design Guidelines as well:

> **Clarity is more important than brevity.**
> Although Swift code can be compact, it is a non-goal to enable the smallest possible code with the fewest characters. Brevity in Swift code, where it occurs, is a side-effect of the strong type system and features that naturally reduce boilerplate.

With that in mind I'd like to leave that particular change out of this PR - but I do appreciate your input on this a lot! Would you mind changing the PR accordingly?

Also added one minor comment on the swift version check.

File: Sources/Validated.swift [hunk 1]
```diff
@@ -1,3 +1,8 @@
+#if swift(>=3.0)
+#else
+    public typealias ErrorProtocol = ErrorType
+#endif
+
```
Comment (Ben-G) [inline Sources/Validated.swift:5]: Nice idea! Curios, would it be possible to do this only with an `if` block?

`#if swift(<3.0)`

Comment (loganwright) [inline Sources/Validated.swift:5]: It gets mad at me when I try to do that 😄  I think it has to be written this way unfortunately.

**Comment by loganmoseley** [comment]
For what it's worth, I prefer the explicit wrapped type.

**Comment by Ben-G** [comment]
@LoganWright just checking in. Would you be OK with updating the PR to leave the explicit argument on the `Validated` type?

**Comment by loganwright** [comment]
@Ben-G Updated to return to explicitly typed syntax, and added support for the 04-12 snapshot which changes some of the rules of external parameters. Mostly this meant adding `validate(_ value:` This allows it to compile on Swift 3, but creates warnings on 2.2. I understand if you'd prefer to wait on merge for warnings to remove, but they shouldn't cause any issues besides the warnings.

**Comment by Ben-G** [comment]
@LoganWright Thanks a lot for the updates 🎉. Given the warnings in Swift 2.2 I'd wait with merging this PR. Since I don't expect too many changes on the codebase until then, this shouldn't be too painful.

**Comment by loganwright** [comment]
Hi @Ben-G 

Again, thanks for the awesome concept!

I wanted to let you know, that we've been using a lot of your original ideas and molded them into something that fits more into how we wanted to use it over at Vapor. 

I made sure to include all of your licenses in all related files, but would love to share some of the original changes back to this repo if you're interested. Take a look at <a href="https://github.com/qutheory/vapor/tree/master/Sources/Vapor/Validation">source</a> and let me know if there's anything you'd like merged back here. I think we took a pretty different approach to some of it, so it might be worth maintaining the fork.

Thanks for sharing this awesome code w/ the community, and let me know if I can help with other features in the future!

\- Logan

**Comment by Ben-G** [comment]
@LoganWright I took a look at how you used the ideas - really cool 😄  I'll take some time to look into detail what of it is generic enough to bring it back into the core library; but definitely great to see how you evolved the ideas!

File: Sources/Validated.swift (hunk 1 above)
Comment (tanner0101) [inline Sources/Validated.swift:5]: You can do `#if !swift(>3.0)`

**Comment by Ben-G** [comment]
@LoganWright Do you happen to be interested in reviving this branch? 😄

**Comment by AndrewSB** [comment]
@LoganWright it would be awesome to get this to a mergeable point 😄

**Comment by Ben-G** [comment]
We now have separate branches for Swift 2.3 and Swift 3.0 support. I'll close this PR since the other changes are a little outdated.

**Pull request closed without merging**